homotopy-graphics/src/path_util.rs

```diff
@@ -0,0 +1,368 @@
+use euclid::{approxeq::ApproxEq, Vector2D};
+use homotopy_common::hash::FastHashMap;
+use homotopy_core::common::Generator;
+use lyon_geom::{CubicBezierSegment, Line, LineSegment};
+use lyon_path::{path::Builder, Event, Path};
+
+use crate::svg::{render::GraphicElement, shape::Point};
+
+pub fn simplify_graphic<const N: usize>(graphic: &[GraphicElement<N>]) -> Vec<GraphicElement<N>> {
+    let mut new_graphic = Vec::with_capacity(graphic.len());
+    let mut point_elements = Vec::new();
+
+    // (depth, gen) -> Vec<(path, start, end)>
+    let mut grouped_wires =
+        FastHashMap::<(usize, Generator), Vec<(Builder, Point, Point)>>::default();
+
+    for element in graphic {
+        match element {
+            GraphicElement::Surface(g, path) => {
+                new_graphic.push(GraphicElement::Surface(*g, simplify_path(path)));
+            }
+            GraphicElement::Wire(g, depth, path, _) => {
+                let entry = grouped_wires.entry((*depth, *g)).or_default();
+
+                let extremes = path_extremes(path).unwrap();
+
+                // The following loop is the core algo wire merging.
+                // The idea is: if we see a segment that shares an endpoint
+                // with a previously seen segment, concatenate them,
+                // and otherwise, just push it as is.
+                //
+                // Then everything is concatenated end-to-end and passed to
+                // the simplifier. The more continuous segments we can catch,
+                // the more the simplifier can cut out!
+                //
+                // We do this by checking whether the extremes of a wire
+                // under consideration match those of any earlier wire.
+                // This gives 4 cases, and we need to be careful with what needs
+                // reversing each time (I think I got this right).
+                // If no matches work, we move to the next seen wire.
+                let mut it = entry.iter_mut();
+                loop {
+                    if let Some((builder, from, to)) = it.next() {
+                        match (
+                            extremes.0.approx_eq(to),
+                            extremes.1.approx_eq(to),
+                            extremes.1.approx_eq(from),
+                            extremes.0.approx_eq(from),
+                        ) {
+                            (true, _, _, _) => {
+                                builder.extend_from_paths(&[path.as_slice()]);
+                                *to = extremes.1;
+                            }
+                            (_, true, _, _) => {
```

Comment:
Could we please drop this case? It feels wrong to reverse paths, and it will definitely cause problems when we add arrows to wires.